### ISSUE TITLE
Improve forward reference extends over definition of value

### ISSUE THREAD

**Issue by REASY** [scala/bug#12382]
Hello, team,

I encountered compilation error on Scala 2.13.4 and it took sometime to understand what's wrong. Here is minimum example:
```scala
import scala.concurrent.{Await, ExecutionContext, Future}

def foo: Future[Int] = {
  val fInt = Future.successful(1)
  val z = for {
    a <- fInt
  } yield a

  implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.Implicits.global
  z
}
```
Compiler says:
```
On line 4: error: forward reference extends over definition of value z
```
![image](https://user-images.githubusercontent.com/5107562/115956508-63b44c00-a527-11eb-87ee-1af25bfa0aaf.png)

Link to code: https://scastie.scala-lang.org/vLXzS5OwSSSKUnJDvtFQxA

The problem is with `implicit val ec: ExecutionContext`, it is defined after it is referenced in for comprehension. Would be great to improve the error message to provide some hints what causes the problem. 

Thanks

**Comment by SethTisue** [comment]
Scala 3: “ec is a forward reference extending over the definition of z”, which is more informative

**Comment by som-snytt** [comment]
Although I just joked that my SO answers have all depreciated to nothing, I remembered that this is a frequent [topic on that forum](https://stackoverflow.com/search?q=%5Bscala%5D+%22forward+reference+extends+over+definition%22).

My answer, as a co-learner in 2012, to [this question](https://stackoverflow.com/questions/13328502/what-does-forward-reference-extends-over-definition-of-value-mean-in-scala/13328614#13328614) suggested that statement order of synthetic code can induce the error. The actual reason in that case was:
```
object Test {
  def main(args: Array[String]) {
    class NotUsed {val xs = args}
    val dummy = false
    // oops, shadows the parameter
    def args = Seq("a","b","c")
  }
}
```
So I suggest that the error message also specify the location of the referenced definition. In the example, the user stares at the code thinking `args` is the args and not some def a thousand LOC later.
